service/dap/server.go

```diff
@@ -325,53 +353,114 @@ func (s *Server) handleRequest(request dap.Message) {
 	jsonmsg, _ := json.Marshal(request)
 	s.log.Debug("[<- from client]", string(jsonmsg))
 
+	if _, ok := request.(dap.RequestMessage); !ok {
+		s.sendInternalErrorResponse(request.GetSeq(), fmt.Sprintf("Unable to process non-request %#v\n", request))
+		return
+	}
+
+	// These requests, can be handeled regardless of whether the targret is running
 	switch request := request.(type) {
-	case *dap.InitializeRequest:
-		// Required
-		s.onInitializeRequest(request)
-	case *dap.LaunchRequest:
-		// Required
-		s.onLaunchRequest(request)
-	case *dap.AttachRequest:
-		// Required
-		s.onAttachRequest(request)
 	case *dap.DisconnectRequest:
 		// Required
 		s.onDisconnectRequest(request)
+		return
+	case *dap.PauseRequest:
+		// Required
+		// TODO: implement this request in V0
+		s.onPauseRequest(request)
+		return
 	case *dap.TerminateRequest:
 		// Optional (capability ‘supportsTerminateRequest‘)
 		// TODO: implement this request in V1
 		s.onTerminateRequest(request)
+		return
 	case *dap.RestartRequest:
 		// Optional (capability ‘supportsRestartRequest’)
 		// TODO: implement this request in V1
 		s.onRestartRequest(request)
-	case *dap.SetBreakpointsRequest:
-		// Required
-		s.onSetBreakpointsRequest(request)
-	case *dap.SetFunctionBreakpointsRequest:
-		// Optional (capability ‘supportsFunctionBreakpoints’)
-		// TODO: implement this request in V1
-		s.onSetFunctionBreakpointsRequest(request)
-	case *dap.SetExceptionBreakpointsRequest:
-		// Optional (capability ‘exceptionBreakpointFilters’)
-		s.onSetExceptionBreakpointsRequest(request)
+		return
+	}
+
+	s.runningMu.Lock()
+	defer s.runningMu.Unlock()
+
+	// Most requests cannot be processed while the debuggee is running.
+	// We have a couple of options for handling these without blocking
+	// the request loop indefinitely when we are in running state.
+	// --1-- Return a dummy response or an error right away.
+	// --2-- Halt execution, process the request, resume execution.
+	// TODO(polina): do this for setting breakpoints
+	// --3-- Handle such requests asynchronously and let them block until
+	// the process stops or terminates (e.g. using a channel and a single
+	// goroutine to preserve the order). This might not be appropriate
+	// for requests such as continue or step because they would skip
+	// the stop, resuming execution right away. Other requests
+	// might not be relevant anymore when the stop is finally reached, and
+	// state changed from the previous snapshot. The user might want to
+	// resume execution before the backlog of buffered request is cleared,
+	// so we would have to either cancel them or delay processing until
+	// the next stop. In addition, the editor itself might block waiting
+	// for these requests to return. We are not aware of any requests
+	// that would benefit from this approach at this time.
+	if s.running {
```

Comment:
@aarzilli the delve itself keeps track of this already. Is it not possible to expose IsRunning instead of keeping track of the state here with a separate set of lock?

Reply:
We could use State() or export isRunning() to avoid getting more state than we need. However, I think we need to lock more than just the bool, but also some related logic, so things cannot change after you flip the status, but before you do all the processing relying on the status being that. I am still revising the code to make sure I get this right. I don't think the current version is doing everything correctly.

Reply:
That would be equivalent to locking s.running when we read it, unlocking, then locking again when we write it, instead of locking the entire block of code that applies logic related to the status we are reading/writing. Is that sufficient? A sync request cannot be interrupted by a new request because it blocks the request loop. No issue there. But an async request can be interrupted by a sync request while running. So then we have the following
- request loop dispatches async request, sets running to true (locking more or less doesn't matter)
- things are running
- another request comes in
- option 1: lock, see that running==true, unlock
- option 2: lock, see that running==true, keep holding the lock
- async request stops running
- sync request still thinks things are running
- option 1:  async sends stopped event, locks, running=false, unlocks, sync sends is-running error (user experience is confusing - why am I getting an error when the session has clearly stopped?)
- option 2: async request tries to get the lock before sending stopped event AND setting running=false, lock is still held by the sync request until it is done sending the error response, so the error and the stopped event arrive in the right order
Or am I missing something?
That said, I have changed my code locally to do this and got new test failures that I am still trying to sort out.

Reply:
From the point of view of a meat and bones user either is going to look like "I pressed a button just as the application was about to stop" and it's all fine. From a mechanical user point of view the observable sequence is either:

```
← send synchronous request
→ receive error "program is running"
→ receive stop message
```

or:

```
← send synchronous request
→ receive stop message
→ receive error "program is running"
```

you want to prevent the latter but you can't because of the delay between the time where the request is sent and the time it is received and acquires the lock. For that to be impossible the client would have to be holding a lock before sending the request.

Reply:
> you want to prevent the latter but you can't because of the delay between the time where the request is sent and the time it is received and acquires the lock. For that to be impossible the client would have to be holding a lock before sending the request.

But if the goroutine that wants to send a stopped event has acquired the running lock, then the request loop that receives the sync request would not be able to process it until the lock is released, and at that point "running" will be false, so it will not need to return an error.

Reply:
I don't think it matters, you are seeing the answer as if the request happened after the stop event, which it didn't.

Reply:
Too late. I already updated the code :) Does this not matter enough not to have to change it back? 
I agree that this is not the biggest issue here though getting a slightly delayed useful response seems better to me than getting a confusing error that's contradicting what is being shown otherwise.

Reply:
Having multiple locks in play makes me uncomfortable and hurts readability and long term maintainability. So, it would be better if we can avoid the lock.